atomic_reactor/plugins/pre_pull_base_image.py

```diff
@@ -109,32 +125,76 @@ def run(self):
+            return None
+        else:
+            return new_image
+
+    def _load_parent_images_digests(self, images_digests):
+        assert isinstance(images_digests, dict)
+        parents_digests = self.workflow.builder.parent_images_digests
+        parents_digests.update_from_dict(images_digests)
+
+    def _collect_image_digests(self, image):
+        parents_digests = self.workflow.builder.parent_images_digests
+
+        if image in parents_digests:
+            # we already have recorded digests for the image, keep it the same
+            # to prevent race conditions at workers
+            return
+
         manifest_list = self._get_manifest_list(image)
-        new_image = None
 
-        if manifest_list:
-            manifest_list_dict = manifest_list.json()
-            arch_digests = {}
-            build_image_digests = {}
-            image_name = image.to_str(tag=False)
+        if not manifest_list:
+            self.log.warning(
+                "Empty manifest list for image %s. Collected image digests will be "
+                "incomplete and may cause race conditions during build", image
+            )
+            return
 
+        manifest_list_dict = manifest_list.json()
+
+        try:
+            arch_to_platform = get_goarch_to_platform_mapping(self.workflow)
+        except KeyError:
+            self.log.warning('Cannot collect platforms digests for parent images '
+                             'because platform descriptors are not defined')
+        else:
             for manifest in manifest_list_dict['manifests']:
                 arch = manifest['platform']['architecture']
-                arch_digests[arch] = image_name + '@' + manifest['digest']
-
-            present_platform = None
-            try:
-                arch_to_platform = get_goarch_to_platform_mapping(self.workflow)
-                for arch, digest in arch_digests.items():
+                try:
                     present_platform = arch_to_platform[arch]
-                    build_image_digests[present_platform] = digest
+                except KeyError:
+                    self.log.warning("Cannot map architecture='{}' to platform".format(arch))
+                    continue
+                else:
+                    digest = manifest['digest']
+                    self.log.info("Collecting digest for image '%s' ('%s'): '%s'",
+                                  image, present_platform, digest)
+                    parents_digests.update_image_digest(
+                        image, present_platform, digest
+                    )
 
-                if platform not in build_image_digests:
-                    new_image = ImageName.parse(build_image_digests[present_platform])
-            except KeyError:
-                self.log.info('Cannot validate available platforms for base image '
-                              'because platform descriptors are not defined')
+    def _get_image_for_different_arch(self, image, platform):
```

Comment:
I don't think I understand the purpose of this.

Reply:
added docstring